brainiak/eventseg/event.py

```diff
@@ -0,0 +1,356 @@
+"""Event segmentation using a Hidden Markov Model
+
+Given an ROI timeseries, this class uses an annealed fitting procedure to
+segment the timeseries into events with stable activity patterns. After
+learning the signature activity pattern of each event, the model can then be
+applied to other datasets to identify a corresponding sequence of events.
+"""
+
+# Authors: Chris Baldassano and Cătălin Iordan (Princeton University)
+
+import numpy as np
+from scipy import stats
+import logging
+from sklearn.base import BaseEstimator
+from sklearn.utils.validation import check_is_fitted, check_array
+
+logger = logging.getLogger(__name__)
+
+__all__ = [
+    "EventSegment",
+]
+
+
+class EventSegment(BaseEstimator):
+    """Class for learning and representing an event segmentation of
+    continuous fMRI data
+
+    Parameters
+    ----------
+    n_events: int
+        Number of segments to learn
+
+    step_var: Callable[[int], float] : default 4 * (0.98 ** (step - 1))
+        The Gaussian variance to use during fitting, as a function of the
+        number of steps. Should decrease slowly over time.
+
+    n_iter: int : default 500
+        Maximum number of steps to run during fitting
+
+    Attributes
+    ----------
+    p_start, p_end: length n_events+1 ndarray
+        initial and final prior distributions over events
+
+    P: n_events+1 by n_events+1 ndarray
+        HMM transition matrix
+
+    ll_ : ndarray with length = number of training datasets
+        Log-likelihood for training datasets over the course of training
+
+    segments_:  list of (time by event) ndarrays
+        Learned (soft) segmentation for training datasets
+
+    event_var_ : float
+        Gaussian variance at the end of learning
+
+    event_pat_ : voxel by event ndarray
+        Learned mean patterns for each event
+    """
+
+    def _default_var_schedule(step):
+        return 4 * (0.98 ** (step - 1))
+
+    def __init__(self, n_events=2,
+                 step_var=_default_var_schedule,
+                 n_iter=500):
+        self.n_events = n_events
+        self.step_var = step_var
+        self.n_iter = n_iter
+
+    def fit(self, X, y=None):
+        """Learn a segmentation on training data
+
+        Fits event patterns and a segmentation to training data. After
+        running this function, the learned event patterns can be used to
+        segment other datasets using find_events
+
+        Parameters
+        ----------
+        X: time by voxel ndarray, or a list of such ndarrays
+            fMRI data to be segmented. If a list is given, then all datasets
+            are segmented simultaneously with the same event patterns
+
+        y: not used (added to comply with BaseEstimator definition)
+
+        Returns
+        -------
+        self: the EventSegment object
+        """
+
+        if type(X) is not list:
+            X = check_array(X)
+            X = [X]
+
+        n_train = len(X)
+        for i in range(n_train):
```

Comment:
I wonder if this can be avoided, e.g., by making copy of X and transpose the copy?
Because X is passed as reference, this line and line 118 which modify X will modify the data X outside the scope of this function, which might be unexpected by the user. For example, if a user wants to fit the data for multiple times, each time trying a different choice of K, (s)he might find the data has been modified after each time of fitting.